games/guess-it/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+                game: Game::new(None, name, bet),
+            }.instantiate()
+        }
+
+        pub fn join(&mut self, mut payment: Bucket) -> (Bucket, Bucket, String) {
+            assert_eq!(self.game.state, State::AcceptingChallenger, "Not accepting challengers!");
+            assert!(self.game.players.len() <= 2, "Too many players!");
+
+            let player: Player = Player::empty();
+            let badge: Bucket = self.badges.take(Decimal::one());
+            let nft_id: NonFungibleKey = badge.get_non_fungible_key();
+            // Secure the funds
+            self.bank.put(payment.take(Decimal::from(self.game.bet as i128)));
+            // Add the player's details
+            self.game.players.insert(nft_id.clone(), player.clone());
+            self.game.state = if self.game.players.len() == 2 { State::MakeGuess } else { State::AcceptingChallenger };
+
+            return (badge, payment, nft_id.to_string());
+        }
+
+        pub fn state(&mut self) -> String {
+            Game::serialize(&self.game)
+        }
+
+        #[auth(badge_ref)]
+        pub fn make_guess(&mut self, guess: u128) -> String {
+            assert_eq!(self.game.state, State::MakeGuess, "Not ready to make a guess!");
+
+            let key = auth.get_non_fungible_key();
+            let player = self.game.players.get_mut(&key).unwrap();
+            player.guess = guess;
+
+            self.game.state = if self.game.has_guessed() { State::WinnerSelection } else { State::MakeGuess };
+
+            if self.game.state == State::WinnerSelection {
+                let response = self.check_winner();
+                if self.game.winner != NonFungibleKey::from(0) {
+                    self.game.state = State::Payout;
+                }
+                format!("Your guess '{}' was accepted\nResponse: {}", guess, response)
+            } else {
+                format!("Your guess '{}' was accepted", guess)
+            }
+        }
+
+        pub fn check_winner(&mut self) -> String {
+            // Pick random number and store it
+            let random_number = (self.game.get_random() % 6) + 1;
+            // determine who was closest without going over
+            self.game.winner = self.game.get_winner(random_number);
+            self.game.last_roll = random_number;
+
+            format!("Dice roll was: '{}' .... winner is: {}",random_number, self.game.winner)
+        }
+
+        #[auth(badge_ref)]
+        pub fn withdraw_funds(&mut self) -> (Bucket, String) {
+            assert_eq!(self.game.state, State::Payout, "It's not time to get paid!");
+            let nft_id: NonFungibleKey = auth.get_non_fungible_key();
+
+            if nft_id == self.game.winner {
+                let amt = self.bank.amount();
+                let payout = self.bank.take(amt);
+
+                (payout, format!("You withdrew {:?} XRD!", amt))
+            }
+            else { panic!("You cannot withdraw funds!") }
```

Comment:
Just a style preference but a simpler way to write is this failing fast:
```
assert!(nft_id == self.game.winner, "You cannot withdraw funds!");

let amt = self.bank.amount();
let payout = self.bank.take(amt);

(payout, format!("You withdrew {:?} XRD!", amt))
```
This way, you only deal with happy path without worrying about alternative flows to panic

Reply:
Thx. Will update next commit